### ISSUE TITLE
TypeError: Cannot read properties of null (reading 'push'). router.push

### ISSUE THREAD

**Issue by The-Code-Monkey** [vercel/next.js#42356]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64
    Binaries:
      Node: 16.15.0
      npm: 8.5.5
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 13.0.2-canary.0
      eslint-config-next: 13.0.2-canary.0
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

In a client side component I keep gettting a type error on push from router.

I am using next/router due to it being client side.

### Expected Behavior

It should redirect like normal.

### Link to reproduction

https://paper-cms.vercel.app/auth/login

### To Reproduce

attempt to log in using 

andy.wilson1234@gmail.com
testing

you will see an error in the console.

code is found here. https://github.com/The-Code-Monkey/PaperCMS/blob/main/app/auth/login/Form.tsx

**Comment by Smirow** [comment]
You should import from `next/navigation` and not `next/router`

```javascript
import { useRouter } from "next/navigation";
```

**Comment by The-Code-Monkey** [comment]
@Smirow I thought you needed next/router when it's client and next/navigation when server. Also im pretty sure I tried it and it did nothing

**Comment by Smirow** [comment]
@The-Code-Monkey see [the docs here](https://beta.nextjs.org/docs/api-reference/use-router).

**Comment by The-Code-Monkey** [comment]
Ah ok I must have been mistaken I will try it out tomorrow

**Comment by The-Code-Monkey** [comment]
Ok i think i've found the issue and its kinda serious.

So using next router does fix it but only in a certain case.

say my directory looks like this

```
app / (admin) / page.tsx
```

router.push('/') doesnt work

but if it was 

```
app / admin / page.tsx
```

router.push('/') sends this in the console.

GET http://localhost:3001/ 404 (Not Found)

Meaning to me that the next/navigation router does not fully understand the layout route folder structure so doesn't try and redirect you. I can fix this for now but i would say thats pretty broken in terms of functionality.

**Comment by jensen** [comment]
Hey @The-Code-Monkey you should try this with `app/(admin)/admin/page.tsx`. It seems as though you are assuming that groups will add more the the pathname, but they do not. 

I expect you are doing an infinite redirect as a result. Your second example actually creates the `/admin/` route.

Notice in my example I have a `app/(dashboard)/dashboard` https://github.com/jensen/supabase-nextjs/tree/main/app/(dashboard)

**Comment by The-Code-Monkey** [comment]
The thing is I don't want the /admin to be a route just a group and when I have had the admin dir as a group I've just been pushing to / and it doesn't work but when I turn admin into a normal URL dir it works. So all I can assume is that the router doesnt understand an app level group

**Comment by The-Code-Monkey** [comment]
I would have assumed that app/(admin)/page.tsx is the same as app/page.tsx as they should be with my understanding of the docs

**Comment by jensen** [comment]
Ok, but under what condition do you `route.push("/")` from inside a `app/(admin)/page.tsx`. That adds the same page to the stack. Why?

**Comment by The-Code-Monkey** [comment]
I am calling the push from app/Auth/login/page.tsx so they aren't the same route.

**Comment by jensen** [comment]
Ok, now I understand what you are doing. The initial description did not make this clear. You are saying that `router.push("/")` from anywhere doesn't work if you wrap your `/` page in a group?

**Comment by The-Code-Monkey** [comment]
Yeah exactly. Sorry I didnt describe that we'll enough, so yeah if a root dir is inside a group it doesn't redirect to it.

**Comment by jensen** [comment]
Ok, I can't reproduce this in my project. My root `page.tsx` is wrapped in a group. I can navigate to it both with `router.push()` and `<Link href>`.

**Comment by Smirow** [comment]
Same as @jensen I have the same pattern, it works fine.

**Comment by The-Code-Monkey** [comment]
Hmm ok not sure that's very strange.

**Comment by The-Code-Monkey** [comment]
Maybe the issue is to do with the fact that I have a layout file at app/layout.tsx and then another in the admin group might try removing that and see if that's causing the router confusion

**Comment by The-Code-Monkey** [comment]
I might try and make a mini repo tomorrow to replicate the issue

**Comment by jensen** [comment]
When I test your app in production, and login you have an infinite loop on your rpc.

![image](https://user-images.githubusercontent.com/14803/199805277-837106d8-fcc1-43d8-837d-32f19af12dd5.png)

**Comment by The-Code-Monkey** [comment]
Oh yeah, I'm aware of that, and I believe the deployed version is using the admin URL Param instead of the group if the redirects work

**Comment by The-Code-Monkey** [comment]
Cheers for your help I will try some stuff in the morning. I guess I might have to wait until I can use fetch client side as most of the logic runs via context. Due to me setting it up to support different DBs.

**Comment by jensen** [comment]
And if you prevent the infinite render, does it actually redirect.

**Comment by jensen** [comment]
> Cheers for your help I will try some stuff in the morning. I guess I might have to wait until I can use fetch client side as most of the logic runs via context. Due to me setting it up to support different DBs.

You can fetch client side just not the way you are doing it, good luck!

**Comment by The-Code-Monkey** [comment]
@jensen Thanks for your help last night.

Not sure its worked though ive taken the table name function out of the Nav component and put it above in its own function but i still get it called 1000x

**Comment by The-Code-Monkey** [comment]
Ah i believe it seems to be because the instance of the supabaseClient within Provider keeps updating meaning a new one is created. Im not sure why but it is.
